js/ai/src/message.ts

```diff
@@ -0,0 +1,131 @@
+/**
+ * Copyright 2024 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { extractJson } from './extract';
+import { MessageData, Part, ToolRequestPart, ToolResponsePart } from './model';
+
+/**
+ * Message represents a single role's contribution to a generation. Each message
+ * can contain multiple parts (for example text and an image), and each generation
+ * can contain multiple messages.
+ */
+export class Message<T = unknown> implements MessageData {
+  role: MessageData['role'];
+  content: Part[];
+  metadata?: Record<string, any>;
+
+  static parseData(
```

Comment:
oh, can you also add
```ts
Message.fromText('foo')
Message.fromMedia('uri')
```

Reply:
hmm... we also need the role... maybe 

```ts
Message.fromText('system', 'foo')
```

Basically I'd like to improve:
```ts
await ai.generate({
  messages: [ Message.fromText('system', 'foo') ]
})
```

Reply:
```ts
await ai.generate({
  messages: [ Message.fromText('system', 'foo') ]
})
```

vs 

```ts
await ai.generate({
  messages: [ { role: 'system'}, content: [{ text: 'foo' }] ]
})
```

not sure anymore. :D

Reply:
feel free to ignore. :)

Reply:
I'm going to ignore just because we actually are already doing some friendly coercing here. This works:

```ts
messages: [{role: "system", content: "talk like a pirate"}]
```